staging/src/k8s.io/apiextensions-apiserver/pkg/controller/apiapproval/apiapproval_controller.go

```diff
@@ -0,0 +1,262 @@
+		crdLister:          crdInformer.Lister(),
+		crdSynced:          crdInformer.Informer().HasSynced,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kube_api_approved_condition_controller"),
+		lastSeenGeneration: map[string]int64{},
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+		DeleteFunc: c.deleteCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+// calculateCondition determines the new KubeAPIApproved condition
+func calculateCondition(crd *apiextensions.CustomResourceDefinition) *apiextensions.CustomResourceDefinitionCondition {
+	if !apihelpers.IsProtectedCommunityGroup(crd.Spec.Group) {
+		return nil
+	}
+
+	switch approvalState := apihelpers.GetAPIApprovalState(crd.Annotations); approvalState {
+	case apihelpers.APIApprovalInvalid:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved, the format is invalid: protected groups must have approval approval with either a URL or a reason starting with \"unapproved\", see https://github.com/kubernetes/enhancements/pull/1111", crd.Name),
+		}
+	case apihelpers.APIApprovalMissing:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved: protected groups must have approval, see https://github.com/kubernetes/enhancements/pull/1111", crd.Name),
+		}
+	case apihelpers.APIApproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionTrue,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v was approved in %v", crd.Name, crd.Annotations[v1beta1.KubeAPIApprovedAnnotation]),
+		}
+	case apihelpers.APIUnapproved:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved: %q", crd.Name, crd.Annotations[v1beta1.KubeAPIApprovedAnnotation]),
+		}
+	default:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionUnknown,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v has an unrecognized approval state %v; coding error", crd.Name, approvalState),
+		}
+	}
+}
+
+func (c *KubeAPIApprovedConditionController) sync(key string) error {
+	inCustomResourceDefinition, err := c.crdLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// avoid repeated calculation for the same generation
```

Comment:
if annotation doesn't change generation, then this will miss recalculating if the annotation changes, right?

Reply:
I agree with the principal that we want to avoid dueling API servers if we change condition calculation in the future, but I think the input should be the annotation value, not the generation